github/migrations_user.go

```diff
@@ -0,0 +1,214 @@
+//
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package github
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+)
+
+// UserMigration represents a GitHub migration (archival).
+type UserMigration struct {
+	ID   *int64  `json:"id,omitempty"`
+	GUID *string `json:"guid,omitempty"`
+	// State is the current state of a migration.
+	// Possible values are:
+	//     "pending" which means the migration hasn't started yet,
+	//     "exporting" which means the migration is in progress,
+	//     "exported" which means the migration finished successfully, or
+	//     "failed" which means the migration failed.
+	State *string `json:"state,omitempty"`
+	// LockRepositories indicates whether repositories are locked (to prevent
+	// manipulation) while migrating data.
+	LockRepositories *bool `json:"lock_repositories,omitempty"`
+	// ExcludeAttachments indicates whether attachments should be excluded from
+	// the migration (to reduce migration archive file size).
+	ExcludeAttachments *bool         `json:"exclude_attachments,omitempty"`
+	URL                *string       `json:"url,omitempty"`
+	CreatedAt          *string       `json:"created_at,omitempty"`
+	UpdatedAt          *string       `json:"updated_at,omitempty"`
+	Repositories       []*Repository `json:"repositories,omitempty"`
+}
+
+func (m UserMigration) String() string {
+	return Stringify(m)
+}
+
+// UserMigrationOptions specifies the optional parameters to Migration methods.
+type UserMigrationOptions struct {
+	// LockRepositories indicates whether repositories should be locked (to prevent
+	// manipulation) while migrating data.
+	LockRepositories bool
+
+	// ExcludeAttachments indicates whether attachments should be excluded from
+	// the migration (to reduce migration archive file size).
+	ExcludeAttachments bool
+}
+
+// startUserMigration represents the body of a StartMigration request.
+type startUserMigration struct {
+	// Repositories is a slice of repository names to migrate.
+	Repositories []string `json:"repositories,omitempty"`
+
+	// LockRepositories indicates whether repositories should be locked (to prevent
+	// manipulation) while migrating data.
+	LockRepositories *bool `json:"lock_repositories,omitempty"`
+
+	// ExcludeAttachments indicates whether attachments should be excluded from
+	// the migration (to reduce migration archive file size).
+	ExcludeAttachments *bool `json:"exclude_attachments,omitempty"`
+}
+
+// StartUserMigration starts the generation of a migration archive.
+// repos is a slice of repository names to migrate.
+//
+// GitHub API docs: https://developer.github.com/v3/migrations/users/#start-a-user-migration
+func (s *MigrationService) StartUserMigration(ctx context.Context, repos []string, opt *UserMigrationOptions) (*UserMigration, *Response, error) {
+	u := "user/migrations"
+
+	body := &startUserMigration{Repositories: repos}
+	if opt != nil {
```

Comment:
Would it be better to use the [`addOptions` function](https://github.com/google/go-github/blob/master/github/github.go#L205) instead?

You can see examples of this being done before across the codebase, including [here](https://github.com/google/go-github/blob/b93adc1c7febaf7dad1bed965e5310a86bc2623e/github/activity_events.go#L152)

Reply:
Looks like addOptions function is used to add query parameters to url.
May be name `UserMigrationOptions` is confusing. Can change `StartUserMigration` methods parameter to accept `startUserMigration` stuct directly, where repos list will be mandatory.

@juliaferraioli Please correct me if I am wrong and also suggest about above approach or any other approach.

Reply:
@sanketpayghan No, you're totally right! I missed that. The `...Options` name does seem misleading.  I think that might be a good change to make. WDYT @gmlewis?

Reply:
I'm thinking that `...Options` makes sense from the perspective of the user of this API because it modifies the behavior of the call despite not going through the internal `addOptions` mechanism.

I've looked through the code, and we do have precedent for using the name `...Options` even when `addOptions` is not involved... for example:

* `Client.Markdown`
* `MigrationService.StartMigration`
* `OrganizationsService.CreateOrgInvitation`
* `ProjectsService.CreateProjectColumn`
* `ProjectsService.MoveProjectColumn`
* `ProjectsService.UpdateProject`
* `ProjectsService.UpdateProjectColumn`
* `PullRequestsService.GetRaw`
* `PullRequestsService.Merge`
* `RepositoriesService.AddCollaborator`
* `RepositoriesService.CreateProject`
* `RepositoriesService.GetArchiveLink`
* `RepositoriesService.GetCommitRaw`
* `SearchService.*`

This was just a cursory glance... I may have got some of these wrong, but I think it is OK to move forward with the current naming scheme, as the typical end-user of this repo probably doesn't care about the internals, but the word `Options` should make sense to them for tweaking the behavior of the endpoint. At least, that's my $0.02. :smile:

Reply:
SGTM!

Reply:
OK, great. Thank you, @juliaferraioli! Do you mind giving this an official LGTM approval so we can go ahead and merge?

Reply:
@gmlewis was just fishing out my laptop :-) Done.